solutionbox/code_free_ml/mltoolbox/code_free_ml/data/trainer/task.py

```diff
@@ -0,0 +1,849 @@
+
+  if keep_target:
+    intermediate_dir = 'intermediate_evaluation_models'
+  else:
+    intermediate_dir = 'intermediate_prediction_models'
+
+  return export_strategy.ExportStrategy(intermediate_dir, export_fn)
+
+
+# TODO(brandondutra) add this to tft, and then remove it.
+def build_csv_transforming_training_input_fn(raw_metadata,
+                                             transform_savedmodel_dir,
+                                             raw_data_file_pattern,
+                                             training_batch_size,
+                                             raw_keys,
+                                             transformed_label_keys,
+                                             convert_scalars_to_vectors=True,
+                                             num_epochs=None,
+                                             randomize_input=False,
+                                             min_after_dequeue=1,
+                                             reader_num_threads=1):
+  """Creates training input_fn that reads raw csv data and applies transforms.
+
+  Args:
+    raw_metadata: a `DatasetMetadata` object describing the raw data.
+    transform_savedmodel_dir: a SavedModel directory produced by tf.Transform
+      embodying a transformation function to be applied to incoming raw data.
+    raw_data_file_pattern: List of files or pattern of file paths containing
+        `Example` records. See `tf.gfile.Glob` for pattern rules.
+    training_batch_size: An int or scalar `Tensor` specifying the batch size to
+      use.
+    raw_keys: List of string keys giving the order in the csv file.
+    transformed_label_keys
+    convert_scalars_to_vectors: Boolean specifying whether this input_fn should
+      convert scalars into 1-d vectors.  This is necessary if the inputs will be
+      used with `FeatureColumn`s as `FeatureColumn`s cannot accept scalar
+      inputs. Default: True.
+
+    num_epochs
+    randomize_input
+    min_after_dequeue
+    reader_num_threads
+    queue_capacity
+
+  Returns:
+    An input_fn suitable for training that reads raw csv training data and
+    applies transforms.
+
+  """
+
+  if not raw_keys:
+    raise ValueError("raw_keys must be set.")
+
+  column_schemas = raw_metadata.schema.column_schemas
+
+  # Check for errors.
+  for k in raw_keys:
+    if k not in column_schemas:
+      raise ValueError("Key %s does not exist in the schema" % k)
+    if not isinstance(column_schemas[k].representation,
+                      dataset_schema.FixedColumnRepresentation):
+      raise ValueError(("CSV files can only support tensors of fixed size"
+                        "which %s is not.") % k)
+    shape = column_schemas[k].tf_shape().as_list()
+    if shape and shape != [1]:
+      # Column is not a scalar-like value. shape == [] or [1] is ok.
+      raise ValueError(("CSV files can only support features that are scalars "
+                        "having shape []. %s has shape %s")
```

Comment:
I don't think this is true unless you extend the definition of csv files. How do you encode the 3d vector (1,2,3) (which is a 1-D tensor) as one csv column?

Maybe "data,1 2 3,data" or "data,1|2|3,data"?  If you do this, you just invented a new file format called "csv+"!